[WIP Not ready for review/merge]Feat: Adding applylib for use by kro reconciler.

barney-s opened the pull request:
Applylib is inspired from:
* kubectl pkg/cmd/apply/applyset.go
* kubebuilder-declarative-pattern/applylib

We are creating a new version of the library that is simpler, self-contained and is purposed built for controllers.

allamand commented:
@barney-s i tried to compile from your branch and to deploy using kro image from this build, and redeploy our solution, but kro fails to create other objects: it managed to create a vpc, but not the other objects like elasticip, subnets...

It should deploy at first this RGD : vpc.kro.run                v1alpha1   Vpc                Active  ["vpc","internetGateway","eip1","eip2","publicRoutetable","publicSubnet1","publicSubnet2","natGateway1","natGateway2","privateRoutetable1","privateRoutetable2","privateSubnet1","privateSubnet2"]  

this is the logs i get
```
+ kro-69fcdb9b8b-vp554 › kro
kro-69fcdb9b8b-vp554 kro 2025-05-23T20:08:33.972Z       ERROR   dynamic-controller      Error syncing item, requeuing with rate limit   {"item": {"NamespacedKey":"argocd/cluster-pre-prod","GVR":{"Group":"kro.run","Version":"v1alpha1","Resource":"eksclusterwithvpcs"}}, "error": "failed to add resource to applyset: failed to synchronize after apply/prune: failed to evaluate dynamic variables: incomplete data: failed evaluating expression vpc.status.privateSubnet2ID: no such key: privateSubnet2ID"}
kro-69fcdb9b8b-vp554 kro 2025-05-23T20:08:34.000Z       ERROR   dynamic-controller      Error syncing item, requeuing with rate limit   {"item": {"NamespacedKey":"argocd/cluster-prod-eu","GVR":{"Group":"kro.run","Version":"v1alpha1","Resource":"eksclusterwithvpcs"}}, "error": "failed to add resource to applyset: failed to synchronize after apply/prune: failed to evaluate dynamic variables: incomplete data: failed evaluating expression vpc.status.privateSubnet2ID: no such key: privateSubnet2ID"}
kro-69fcdb9b8b-vp554 kro 2025-05-23T20:08:34.028Z       ERROR   dynamic-controller      Error syncing item, requeuing with rate limit   {"item": {"NamespacedKey":"argocd/cluster-prod-us","GVR":{"Group":"kro.run","Version":"v1alpha1","Resource":"eksclusterwithvpcs"}}, "error": "failed to add resource to applyset: failed to synchronize after apply/prune: failed to evaluate dynamic variables: incomplete data: failed evaluating expression vpc.status.privateSubnet1ID: no such key: privateSubnet1ID"}
kro-69fcdb9b8b-vp554 kro 2025-05-23T20:08:34.064Z       ERROR   dynamic-controller      Error syncing item, requeuing with rate limit   {"item": {"NamespacedKey":"argocd/cluster-test","GVR":{"Group":"kro.run","Version":"v1alpha1","Resource":"eksclusterwithvpcs"}}, "error": "failed to add resource to applyset: failed to synchronize after apply/prune: failed to evaluate dynamic variables: incomplete data: failed evaluating expression vpc.status.vpcID: no such key: vpcID"}
kro-69fcdb9b8b-vp554 kro 2025-05-23T20:08:34.204Z       ERROR   dynamic-controller      Error syncing item, requeuing with rate limit   {"item": {"NamespacedKey":"argocd/cluster-pre-prod","GVR":{"Group":"kro.run","Version":"v1alpha1","Resource":"eksclusterwithvpcs"}}, "error": "failed to add resource to applyset: failed to synchronize after apply/prune: failed to evaluate dynamic variables: incomplete data: failed evaluating expression vpc.status.privateSubnet2ID: no such key: privateSubnet2ID"}
kro-69fcdb9b8b-vp554 kro 2025-05-23T20:08:34.237Z       ERROR   dynamic-controller      Error syncing item, requeuing with rate limit   {"item": {"NamespacedKey":"argocd/cluster-prod-eu","GVR":{"Group":"kro.run","Version":"v1alpha1","Resource":"eksclusterwithvpcs"}}, "error": "failed to add resource to applyset: failed to synchronize after apply/prune: failed to evaluate dynamic variables: incomplete data: failed evaluating expression vpc.status.privateSubnet1ID: no such key: privateSubnet1ID"}
```

allamand commented:
If I revert to the old kro version i was using, then the spec is correctly created in the generated vpc crd:

```
kind: Vpc
metadata:
  annotations:
    applyset.kubernetes.io/additional-namespaces: cluster-test
    applyset.kubernetes.io/contains-group-kinds: ',ElasticIPAddress.ec2.services.k8s.aws,InternetGateway.ec2.services.k8s.aws,VPC.ec2.services.k8s.aws'
    applyset.kubernetes.io/tooling: kro/v0.3.1-0.20250523032018-f3e42082fdbf+dirty
    argocd.argoproj.io/tracking-id: clusters:kro.run/Vpc:cluster-test/cluster-test
  creationTimestamp: "2025-05-26T08:15:22Z"
  finalizers:
  - kro.run/finalizer
  generation: 3
  labels:
    app.kubernetes.io/instance: cluster-test
    applyset.kubernetes.io/id: applyset-F5j7EzsvRC4sAiP7GLhO4SfBcQeKPscHG9KbnRiXlZE-v1
    kro.run/controller-pod-id: kro-pod
    kro.run/instance-id: 5ab48226-5910-4c13-8708-a9e7271bfed5
    kro.run/instance-name: cluster-test
    kro.run/instance-namespace: argocd
    kro.run/kro-version: 0.2.1
    kro.run/owned: "true"
    kro.run/resource-graph-definition-id: a1649b15-fdca-4152-bda6-704801d74344
    kro.run/resource-graph-definition-name: eksclusterwithvpc.kro.run
  name: cluster-test
  namespace: cluster-test
  resourceVersion: "2032005"
  uid: 4f879ece-8930-4851-aa59-c9a76482cb39
spec:
  cidr:
    privateSubnet1Cidr: 10.0.11.0/24
    privateSubnet2Cidr: 10.0.12.0/24
    publicSubnet1Cidr: 10.0.1.0/24
    publicSubnet2Cidr: 10.0.2.0/24
    vpcCidr: 10.0.0.0/16
  name: cluster-test
  region: eu-central-1
```

barney-s commented:
> @barney-s I have an error with the new build, I create an RGD (eksclusterwithvpc) that create another RGD (vpc). the Issue is that the vpc object created is missing the spec part when using this PR version:

The issue was with fieldmanager being the same across multiple reconcile paths.
Separated out the field managers and fixed a bug where the parent reconciler (eksclusterwithvpc, the one that creates VPC) and the child reconciler (VPC instance) both were overwriting labels in the VPC instance. 

Please retry with latest version of the PR. 

Was able to reproduce the issue with a simplified version of RGD:

```
apiVersion: kro.run/v1alpha1
kind: ResourceGraphDefinition
metadata:
  name: parent-rgd.kro.run
spec:
  schema:
    apiVersion: v1alpha1
    kind: Parent
    spec:
      name: string
      cidr:
        public: string | default="10.0.3.0/24"
  resources:
  - id: child
    template:
      apiVersion: kro.run/v1alpha1
      kind: Child
      metadata:
        name: ${schema.spec.name}
      spec:
        name: ${schema.spec.name}
        cidr:
          public: ${schema.spec.cidr.public}
---
apiVersion: kro.run/v1alpha1
kind: ResourceGraphDefinition
metadata:
  name: child-rgd.kro.run
spec:
  schema:
    apiVersion: v1alpha1
    kind: Child
    spec:
      name: string
      cidr:
        public: string | default="10.0.1.0/24"
  resources: []

```

Instance:
```
apiVersion: kro.run/v1alpha1
kind: Parent
metadata:
  name: p1
spec:
  name: p1
  cidr:
    public: 10.0.8.0/24
```

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+	DryRun(ctx context.Context, prune bool) (*ApplyResult, error)
+}
+
+type ApplySetConfig struct {
```
done

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+
+// NewApplySet creates a new ApplySet
+// parent object is expected to be the current one existing in the cluster
+func NewApplySet(
```
done

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+	dynamicClient dynamic.Interface,
+	config ApplySetConfig,
+) (ApplySet, error) {
+	force := true
```
removed the variable

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+	callbacks
+}
+
+func (t ApplySetTooling) String() string {
```
agreed

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+	"k8s.io/client-go/dynamic"
+)
+
+type ApplySetTooling struct {
```
renamed to Tooling

barney-s commented on applylib/applyset/const.go:
```diff
@@ -0,0 +1,68 @@
+	// When present, the value of this annotation must be a comma separated list of the group-resources,
+	// in the fully-qualified name format, i.e. <resourcename>.<group>.
+	// Example value: "certificates.cert-manager.io,configmaps,deployments.apps,secrets,services"
+	DeprecatedApplySetGRsAnnotation = "applyset.kubernetes.io/contains-group-resources"
```
removed.

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+	gvk := obj.GroupVersionKind()
+	switch restMapping.Scope.Name() {
+	case meta.RESTScopeNameNamespace:
+		// TODO (barney-s): Should we allow empty namespace for namespace-scoped objects?
```
Treating this as a library, wanted to be defensive. 

Added the check back.

barney-s replied:
Effectively we allow namespace scoped resources to be passed without namespace. So adding the check is effectively short-circuiting the reconcile loop with an error. 

So will leave this check commented for now.

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+type ApplySet interface {
+	Add(ctx context.Context, obj ApplyableObject) error
+	Apply(ctx context.Context, prune bool) (*ApplyResult, error)
+	DryRun(ctx context.Context, prune bool) (*ApplyResult, error)
```
Latter. 

When we apply, we apply the objects from the Set.
Then use the parent object annotations (GVK, N, label) to get the set of objects in cluster from last iteration.
Compare it with the current set and detect objects that we last applied and not in current set. Then we prune those. 
At the end of pruning we update the parent object annotations with the current set => GVK, N

barney-s commented on applylib/applyset/result.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+type ApplyResult struct {
+	Desired        int
```
renamed to DesiredCount

barney-s commented on applylib/applyset/tracker.go:
```diff
@@ -0,0 +1,126 @@
+	clientIDs map[string]bool
+}
+
+func (a *ApplyableObject) Json() []byte {
```
No and removed the .marshalled as well.

barney-s commented on applylib/applyset/tracker.go:
```diff
@@ -0,0 +1,126 @@
+	t.clientIDs[obj.ID] = true
+
+	// Ensure the object is marshallable
+	if j, err := json.Marshal(obj); err != nil {
```
removed.

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Applylib is inspired from:
```
added a link to the KEP in the code.

barney-s commented on applylib/applyset/result.go:
```diff
@@ -0,0 +1,113 @@
+	PrunedObjects  []PrunedObject
+}
+
+func (a *ApplyResult) PruneErrors() error {
```
Good Q. 
Dont think we need 2 separate errors. 
Replacing with single one.

barney-s commented on applylib/applyset/tracker.go:
```diff
@@ -0,0 +1,126 @@
+	objects []ApplyableObject
+
+	// serverIDs is a map of object key to object.
+	serverIDs map[k8sObjectKey]bool
```
i guess i was using sets.Set there since i did not care about if an obj is already present. I keep adding to the set and lettting sets.Set dedupe for me. 

Here i care about if a key already exists to error out. 
Can be achieved with sets.Set.Has(). But i thought this may be more straight forward.

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+		mapping := pruneObject.Mapping
+
+		// TODO: Why is it always using namesapce ?
+		po := PrunedObject{
```
i see why i did this. 
We have a beforePrune callback. That requires the PrunedObject{}
Hence we are constructing it here. And updating the result of prune (err) in the prunedobject before adding it to the result set. 

For apply we have the afterapply callback. Hence we could use the helper function to do it.

barney-s replied:
I ended up adding a helper fn and refactoring the code a bit.

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+}
+
+// NewApplySet creates a new ApplySet
+// parent object is expected to be the current one existing in the cluster
```
Is this comment still relevant ?

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+}
+
+type serverOptions struct {
+	// patchOptions holds the options used when applying, in particular the fieldManager
```
fixed

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+		toolingID:           config.ToolingID,
+		toolLabels:          config.ToolLabels,
+		desired:             NewTracker(),
+		desiredRestMappings: make(map[schema.GroupKind]*meta.RESTMapping),
```
done

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+	// client is the dynamic kubernetes client used to apply objects to the k8s cluster.
+	dynamicClient dynamic.Interface
+
+	// ParentClient is the controller runtime client used to apply parent.
```
done

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+}
+
+type clientSet struct {
+	// client is the dynamic kubernetes client used to apply objects to the k8s cluster.
```
done

barney-s commented on applylib/applyset/prune.go:
```diff
@@ -0,0 +1,173 @@
+
+// PruneObject is an apiserver object that should be deleted as part of prune.
+type PruneObject struct {
+	Name      string
```
ah we had it because i was using runtime.Object. Switched to unstructured and removed Name, Namespace.

barney-s commented on applylib/applyset/prune.go:
```diff
@@ -0,0 +1,173 @@
+	var allObjects []PruneObject
+	for _, task := range tasks {
+		if task.err != nil {
+			return nil, task.err
```
Makes sense. Switched to errgroup.

barney-s commented on applylib/applyset/tracker.go:
```diff
@@ -0,0 +1,126 @@
+	Decorate    bool
+
+	// lastReadRevision is the revision of the object that was last read from the cluster.
+	lastReadRevision string
```
So as part of applyset.Add() we read from cluster (if found) and record the `lastReadRevision`. And later as part of applyset.Apply() we set the `LastApplied` for the `AppliedObject`. It is possible that there was no change as part of the apply and it is a no-op server side. So we check if `lastReadRevision` (before we apply) and the `LastApplied.GetResourceVersion()` (after apply) has changed  to determine if there was a cluster mutation.

barney-s commented on pkg/controller/instance/controller_reconcile.go:
```diff
@@ -446,16 +421,26 @@ func (igr *instanceGraphReconciler) setManaged(ctx context.Context, obj *unstruc
 	updated, err := igr.client.Resource(igr.gvr).
 		Namespace(obj.GetNamespace()).
-		Update(ctx, copy, metav1.UpdateOptions{})
+		Apply(ctx, instancePatch.GetName(), instancePatch, metav1.ApplyOptions{FieldManager: "kro-managed", Force: false})
```
done. more cleanup for label namings.

barney-s commented:
@matthchr - Thanks for the review comments. I have addressed most of them.  Some of them i have resolved (when i made code changes). Things that required clarifications i have responded to the comments but not resolved them. If you feel convinced with the response, please resolve those comments. 

The only comment i have not yet worked on is the `desiredNamespaces` comment. I will spend some time on it and get back.

Meanwhile please take another pass and let me know. Thanks.

barney-s commented on applylib/applyset/tracker.go:
```diff
@@ -0,0 +1,126 @@
+	t.serverIDs[objectKey] = true
+
+	// client side uniqueness check
+	if _, found := t.clientIDs[obj.ID]; found {
```
Add a TODO to see if clientID is needed at applylib level. Can we let kro handle GVKNN to ID mapping ?

barney-s replied:
added a TODO. resolving the comment here.

barney-s commented on pkg/controller/instance/controller_status.go:
```diff
@@ -100,9 +101,19 @@ func (igr *instanceGraphReconciler) patchInstanceStatus(ctx context.Context, sta
-	_, err := igr.client.Resource(igr.gvr).
-		Namespace(instance.GetNamespace()).
-		UpdateStatus(ctx, instance, metav1.UpdateOptions{})
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
TODO: Create a separate issue to handle defer() status update. It is possible for reconcile to be complete and the status is not updated in `func (igr *instanceGraphReconciler) handleReconciliation()`

barney-s commented on applylib/applyset/applyset.go:
```diff
@@ -0,0 +1,521 @@
+	}
+	nsList := a.desiredNamespaces.UnsortedList()
+	sort.Strings(nsList)
+	annotations[ApplySetAdditionalNamespacesAnnotation] = strings.Join(nsList, ",")
```
Looks like a bug. I have fixed it in a separate commit for you to review. PTAL.

barney-s commented on applylib/applyset/prune.go:
```diff
@@ -0,0 +1,173 @@
+	for _, restMapping := range a.desiredRestMappings {
+		switch restMapping.Scope.Name() {
+		case meta.RESTScopeNameNamespace:
+			for _, namespace := range a.desiredNamespaces.UnsortedList() {
```
Fixed in a separate commit. PTAL

jakobmoellerdev commented on test/e2e/chainsaw/check-multi-resource-rgd/instance-create-assert.yaml:
```diff
@@ -6,7 +6,7 @@ metadata:
   labels:
     kro.run/owned: "true"
-    kro.run/resource-graph-definition-name: check-multi-resource-rgd
+    kro.run/source-resource-graph-definition-name: check-multi-resource-rgd
```
this looks a bit off, is this a breaking change? maybe lacking about the understanding of these labels is my issue here, feel free to resolve if non-issue

barney-s replied:
When testing i found the case where we may have recursive RGDs. RGD1 having RGD2 instance as part of its resources. In such a case, RGD2's instance created by RGD1 would have 2 sets of labels. 

The source one which points to the RGD2 as its source definition.
The regular one which points to RGD1 instance as its creator.

jakobmoellerdev commented on test/e2e/chainsaw/check-rgd-deletion/chainsaw-test.yaml:
```diff
@@ -42,7 +42,16 @@ spec:
+    finally:
+    - description: sleep operation
+      sleep:
+        duration: 60s # wait for the RGD to be deleted. Running the test with 5s is flaky (passes locally but fails in CI)
```
60s sounds like a lot, maybe can be reduced?

jakobmoellerdev commented on pkg/controller/instance/controller_reconcile.go:
```diff
@@ -109,6 +129,45 @@ func (igr *instanceGraphReconciler) handleReconciliation(ctx context.Context, re
+	obj applyset.ApplyableObject,
+	observed *unstructured.Unstructured,
+) error {
+	if observed != nil {
```
nit: switch if to `== nil` and put logic out of if

jakobmoellerdev commented on go.mod:
```diff
@@ -14,17 +14,20 @@ require (
 	k8s.io/apimachinery v0.31.0
 	k8s.io/apiserver v0.31.0
 	k8s.io/client-go v0.31.0
+	k8s.io/klog/v2 v2.130.1
```
where do we use klog directly right now, I thought we only use logrus as our lib?

barney-s replied:
its removed now in the latest.

jakobmoellerdev commented on pkg/controller/instance/controller_reconcile.go:
```diff
@@ -123,193 +182,102 @@ func (igr *instanceGraphReconciler) reconcileInstance(ctx context.Context) error
-		if _, err := igr.runtime.Synchronize(); err != nil {
-			return fmt.Errorf("failed to synchronize reconciling resource %s: %w", resourceID, err)
-		}
+	// TODO (barney-s):
```
I think this todo can be removed after https://github.com/kro-run/kro/pull/561/files#r2198966036

barney-s replied:
ah yes. removed.

jakobmoellerdev commented on pkg/metadata/labels.go:
```diff
@@ -41,10 +41,17 @@ const (
 	ResourceGraphDefinitionVersionLabel   = LabelKROPrefix + "resource-graph-definition-version"
+
+	// Source Identifier Labels. These are set by the instance reconciler that reconciles the resource.
+	SourceResourceGraphDefinitionIDLabel        = LabelKROPrefix + "source-resource-graph-definition-id"
```
Maybe stupid question: in which case can we have differing labels here? I.E. is there a case in which we have a differing ResourceGraphDefinitionIDLabel and SourceResourceGraphDefinitionIDLabel

barney-s replied:
When we have a RGD2's instance as part of another RGD1, we want to inject labels that can identify the creator of the instance (RGD1) as well as the reconciler of that instance (RGD2)

jakobmoellerdev commented on pkg/controller/instance/controller_reconcile.go:
```diff
@@ -446,16 +426,26 @@ func (igr *instanceGraphReconciler) setManaged(ctx context.Context, obj *unstruc
 	updated, err := igr.client.Resource(igr.gvr).
 		Namespace(obj.GetNamespace()).
-		Update(ctx, copy, metav1.UpdateOptions{})
+		Apply(ctx, instancePatch.GetName(), instancePatch, metav1.ApplyOptions{FieldManager: FieldManagerForLabeler, Force: false})
```
a force false would lead to errors if we have other field managers on this. Since we are in finalize Deletion, is forcing here a valid case?

barney-s replied:
true. i fail to recollect why i had false here specifically. reverting to true. If we do indeed discover why we need a false later we can add a comment and revert it. For now true makes sense for the force flag.

jakobmoellerdev reviewed (commented):
Nice change! Just went over the PR as requested in the Community Call and called out what looked off to me. Feel free to ignore/resolve if you think its irrelevant

barney-s commented on Makefile:
```diff
@@ -286,7 +293,7 @@ cli:
 .PHONY: test-e2e
 test-e2e: chainsaw ## Run e2e tests
-	$(CHAINSAW) test ./test/e2e/chainsaw
+	$(CHAINSAW) test ./test/e2e/chainsaw/check-rgd-deletion
```
ah indeed. thanks for flagging this,

jakobmoellerdev commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,543 @@
+				return results, fmt.Errorf("error from before-prune callback: %w", err)
+			}
+		}
+		err := a.dynamicClient.Resource(mapping.Resource).Namespace(namespace).Delete(ctx, name, options)
```
does this need something like

```
if mapping.Scope.Name() == meta.RESTScopeNameNamespace {
		err := a.dynamicClient.Resource(mapping.Resource).Namespace(namespace).Delete(ctx, name, options)
	} else {
		err := a.dynamicClient.Resource(mapping.Resource).Delete(ctx, name, options)
}
```
if this doesnt then the code in `New` above also doesnt need it I think

barney-s replied:
added the recommended code to check namespacescope

jakobmoellerdev commented on pkg/controller/instance/controller.go:
```diff
@@ -154,8 +161,9 @@ func (c *Controller) Reconcile(ctx context.Context, req ctrl.Request) error {
 		log:                         log,
 		gvr:                         c.gvr,
 		client:                      executionClient,
+		restMapper:                  c.clientSet.RESTMapper(),
```
Im wondering about this. I believe the RESTMapper from the client set may be wrong? im wondering if the correct client for resolving mappings here should be the stuff behind execution client?
will have to double check this again!

barney-s replied:
Should we revert to the old state where we dont include restMapper as part of clientset ?

jakobmoellerdev commented on pkg/controller/instance/controller_reconcile.go:
```diff
@@ -123,193 +184,100 @@ func (igr *instanceGraphReconciler) reconcileInstance(ctx context.Context) error
-// labels and managed state.
-func (igr *instanceGraphReconciler) setupInstance(ctx context.Context, instance *unstructured.Unstructured) error {
-	patched, err := igr.setManaged(ctx, instance, instance.GetUID())
+	aset, err := applyset.New(instance, igr.restMapper, igr.client, config)
```
nit: im wondering if you need to pass in instance here because of https://github.com/kro-run/kro/pull/561/files#diff-3e23972797bd9d87e2581d861789fbe1816800bf43794f45ae4d46f7a304e790R115 since you have the gvr with igr.gvr, you may want to use that as a parent instead?

barney-s replied:
Extracting  PartialObjectMetadata in New() now. We need the parent name and namespace as well.

jakobmoellerdev commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,543 @@
+	parentPatch.SetAnnotations(annotations)
+
+	options := metav1.ApplyOptions{
+		FieldManager: a.fieldManager + "parent-labeller",
```
is the field-manager set with a prefix already? otherwise this manager will look very strange

barney-s replied:
done.

jakobmoellerdev commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,543 @@
+	// update parent in the cluster.
+	if !reflect.DeepEqual(original.GetLabels(), parentPatch.GetLabels()) ||
+		!reflect.DeepEqual(original.GetAnnotations(), parentPatch.GetAnnotations()) {
+		if _, err := a.parentClient.Apply(ctx, a.parent.GetName(), parentPatch, options); err != nil {
```
this is an interesting one. our parent is our instance (https://github.com/kro-run/kro/pull/561/files#diff-b51fa143abc50a3ef21525703711d6ab24b80395539ca3a6a9a4d3da24050435R196), and we watch the instance in our controller as well AFAIU. Will this accidentally trigger a reconciliation if we don't filter for the annotations in our event handler? or potentially cause an early requeue?

barney-s replied:
Naively i think it should trigger a watch change. How do you filter for annotations ? We dont usually know what field changed right ? 

Maybe it is ok. This happends during create and should "settle down" into a stable state with new reconciles being no-op (no change). 

Thoughts ?

jakobmoellerdev replied:
Actually I agree, the stable state and eventual consistency should win here. +1

jakobmoellerdev reviewed (commented):
This is looking awesome 🥳! Most of my comments are just `nit:` suggestions—feel free to ignore if you don’t think they add value.

A main concern from the Community Call was that the applyset implementation currently relies on side-effects from callbacks (e.g., for synchronization), which was hard to understand. I debugged a bit manually, I think this is acceptable TBH. I’m not sure there’s a better way, but it’s something we should keep an eye on.

Another point: where the apply prune set is stored. In kubectl, a dedicated object (secret/configmap) is used. Currently, We’d reuse the parent object (the instance) here. If so, should we worry about object storage limits? In kubectl, applysets usually handle only 10s of resources because of annotation size limits, and we may hit similar constraints in KRO. Im interested in your thoughts on this!

Overall, my concerns aren’t with the implementation (you did an amazing job 🙌) but more around how we best keep maintaining this (its a semi-fork of kubectl so imo it needs a complete test suited).  I think you mentioned contributing tests in a separate PR— the apply set might be something we just reference there. For me its acceptable due to size constraints.

barney-s commented on pkg/controller/instance/controller_status.go:
```diff
@@ -100,9 +101,19 @@ func (igr *instanceGraphReconciler) patchInstanceStatus(ctx context.Context, sta
-		Namespace(instance.GetNamespace()).
-		UpdateStatus(ctx, instance, metav1.UpdateOptions{})
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		instance, err := igr.client.Resource(igr.gvr).
```
We discussed this in some meeting and probably did not capture it well enough (as a bug or comment). The reason we are doing this is the method is called in a defer path. There is no way to return an error to the reconciler to requeue from that method. Alternative is to change the top level reconciler implementation to remove the defer.

barney-s replied:
added a comment to that effect.

barney-s commented on pkg/metadata/finalizers.go:
```diff
@@ -44,32 +44,20 @@ func HasResourceGraphDefinitionFinalizer(obj metav1.Object) bool {
-
-	if found {
+	finalizers := obj.GetFinalizers()
+	if containsString(finalizers, kroFinalizer) {
```
done

barney-s commented on pkg/applyset/result.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+func (a *ApplyResult) pruneErrors() error {
+	errorsSeen := []error{}
```
TIL. Thanks. Made the suggested change.

barney-s commented on pkg/applyset/result.go:
```diff
@@ -0,0 +1,130 @@
+	return errors.Join(errorsSeen...)
+}
+
+func (a *ApplyResult) applyErrors() error {
```
done.

barney-s commented on pkg/applyset/prune.go:
```diff
@@ -0,0 +1,162 @@
+
+	var allObjects []PruneObject
+	for _, task := range tasks {
+		allObjects = append(allObjects, task.results...)
```
We run a task per GK. So dont expect resources to overlap. If due to a bug (none i see for now) if we have duplicate GKs, the delete of duplicate objects would be recorded as an error in the prune-results.

barney-s commented on pkg/applyset/prune.go:
```diff
@@ -0,0 +1,162 @@
+	"k8s.io/client-go/dynamic"
+)
+
+type ApplySetDeleteOptions struct {
```
its not even being used currently. removed it.

barney-s commented on pkg/applyset/prune.go:
```diff
@@ -0,0 +1,162 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package applyset
```
Added a comment saying we have copied over code from kubectl and the path.

barney-s commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,543 @@
+	// Parent provides the necessary methods to determine a
+	// ApplySet parent object, which can be used to find out all the on-track
+	// deployment manifests.
+	parent *unstructured.Unstructured
```
Using PartialObjectMetadata

barney-s commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,543 @@
+// updateParentLabelsAndAnnotations updates the parent labels and annotations.
+func (a *applySet) updateParentLabelsAndAnnotations(ctx context.Context, mode applySetUpdateMode) (sets.Set[string],
+	sets.Set[string], error) {
+	original, err := meta.Accessor(a.parent.DeepCopyObject())
```
correct. removed the deepcopy

barney-s commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,543 @@
+			}
+		}
+
+		// TODO: Implement health computation after apply
```
We can remove this comment. We will defer this for now. KRO reconcile loop may be another place to compute health.

jakobmoellerdev commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,543 @@
+		a.parent.GetKind(),
+		a.parent.GroupVersionKind().Group,
+	}, ApplySetIDPartDelimiter)
+	hashed := sha256.Sum256([]byte(unencoded))
```
Hmm interesting. I think UID is more deterministic than Name+Namespace right? Because if you have an object recreated it will have the same Name/Namespace but different UID right? and since the parent is the object we store the apply set on, the UID would be the canonical source right?

barney-s commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,543 @@
+		namespace := pruneObjects[i].GetNamespace()
+		mapping := pruneObjects[i].Mapping
+
+		// TODO: Why is it always using namesapce ?
```
added code to check for namespace-scoped.

barney-s commented:
> This is looking awesome 🥳! Most of my comments are just `nit:` suggestions—feel free to ignore if you don’t think they add value.

Thanks for the detailed comments. Addressed most of them. PTAL again. 
 
> A main concern from the Community Call was that the applyset implementation currently relies on side-effects from callbacks (e.g., for synchronization), which was hard to understand. I debugged a bit manually, I think this is acceptable TBH. I’m not sure there’s a better way, but it’s something we should keep an eye on.

Callbacks was a simpler way of dealing with the way KRO handles Resource Graph. Since callbacks are optional, Applyset would work for standard reconciler and cli use-cases.
 
> Another point: where the apply prune set is stored. In kubectl, a dedicated object (secret/configmap) is used. Currently, We’d reuse the parent object (the instance) here. If so, should we worry about object storage limits? In kubectl, applysets usually handle only 10s of resources because of annotation size limits, and we may hit similar constraints in KRO. Im interested in your thoughts on this!


I believe in kubectl we can pass a custom parent object:  `--applyset=[resource.version.group/]name`

https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/3659-kubectl-apply-prune/README.md#kubectl-commands-and-flags

Applyset design was designed to use minimal metadata (GVK, NS) to track child objects along with labels. 
We believe this is the least amount of metadata that needs to be tracked. Alternative is to scan all GVKs and all namespaces which may result in performance issues. 

Re: storage limits, we think this should be ok to begin with. We may run into scaling issues with reconciler before hitting object size limits. If we do hit limits, we can switch to an alternate mode where no GVK, NS are tracked. 
 
> Overall, my concerns aren’t with the implementation (you did an amazing job 🙌) but more around how we best keep maintaining this (its a semi-fork of kubectl so imo it needs a complete test suited). I think you mentioned contributing tests in a separate PR— the apply set might be something we just reference there. For me its acceptable due to size constraints.

The goal is to upstream to cli-utils or kubectl. We will work with them on this. Alternate location is controller runtime.

jakobmoellerdev reviewed (approved):
This has been awesome work done by you. Thanks for taking the time to walk with me through all the comments. LGTM 🥳 interested to see what other maintainers have to say on this

jakobmoellerdev commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,543 @@
+	return nil
+}
+
+func (a *applySet) validateAndCacheRestMapping(obj ApplyableObject) (*meta.RESTMapping, error) {
```
I think we have controller-runtime as hard dependency anyways but if you want to contribute even further upstream (k8s core repos) this is totally okay for me 👍

barney-s replied:
i meant controller runtime dependency for just pkg/applyset. a cursory look of imports in the folder shows no controller-runtime. Hope there is no indirect dependency.

jakobmoellerdev replied:
Right, I was referring to our global project based go.mod (https://github.com/kro-run/kro/blob/ea6885a9e28ddb444ddc3b315d37baf139c5ca8d/go.mod#L27) on which we have a controller runtime dependency. But knowing that you wanna contribute applyset upstream this is fine for me

n3wscott commented on pkg/applyset/prune.go:
```diff
@@ -0,0 +1,163 @@
+// limitations under the License.
+
+// Applylib has code copied over from:
+//   - kubectl pkg/cmd/apply/applyset.go
```
I believe we also need to bring over the copyright headers from the forked code in each file to properly attribute the code.

barney-s replied:
Added  `Copyright 2023 The Kubernetes Authors.`

Rest of the license is common.

n3wscott commented on pkg/applyset/prune.go:
```diff
@@ -0,0 +1,163 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Applylib has code copied over from:
```
This explication should be expanded and moved to a docs.go file to better explain this package, why it was forked, and when to unfork it.

barney-s replied:
adding a docs.go

n3wscott commented on pkg/metadata/finalizers.go:
```diff
@@ -44,32 +45,18 @@ func HasResourceGraphDefinitionFinalizer(obj metav1.Object) bool {
-	}
-
-	if !found || !containsString(finalizers, kroFinalizer) {
+	finalizers := obj.GetFinalizers()
```
this finalizer code looks unrelated to the PR, better to pull it to a seperate PR for better visibility.

barney-s replied:
https://github.com/kro-run/kro/pull/625

n3wscott commented on pkg/controller/resourcegraphdefinition/controller_reconcile.go:
```diff
@@ -87,9 +92,24 @@ func (r *ResourceGraphDefinitionReconciler) reconcileResourceGraphDefinition(ctx
+) (metadata.Labeler, metadata.Labeler, error) {
 	rgLabeler := metadata.NewResourceGraphDefinitionLabeler(rgd)
-	return r.metadataLabeler.Merge(rgLabeler)
+	rgSourceLabeler := metadata.NewResourceGraphDefinitionSourceLabeler(rgd)
```
is this required to crate applyset for use by instance reconciler? Or are the labels new? Is this needed in this PR?

barney-s replied:
We are introducing a new label. We need this to deal with a bug in KRO where if we have nested RGD (RGD1 creating instance of RGD2), the reconciler of RGD1 and RGD2 both operate on the same label. We added a source label to differentiate it. Similar comments by @jakobmoellerdev that were discussed.

n3wscott commented on pkg/controller/instance/controller_reconcile.go:
```diff
@@ -20,13 +20,16 @@ import (
-	"github.com/kro-run/kro/pkg/controller/instance/delta"
+	"sigs.k8s.io/release-utils/version"
+
+	"github.com/kro-run/kro/pkg/applyset"
```
ahhhhh here it is used. Sorry this was too large to render when I looked the first time

n3wscott commented on pkg/controller/instance/controller_reconcile.go:
```diff
@@ -44,6 +47,16 @@ const (
 	ResourceStateWaitingForReadiness = "WAITING_FOR_READINESS"
 	ResourceStateUpdating            = "UPDATING"
+
+	FieldManagerForApplyset = "kro.run/applyset"
```
the applyset code defines labels, are those not needed?

barney-s replied:
This is for field manager in apply. Labels are orthogonal.

n3wscott commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,546 @@
+}
+
+type Set interface {
+	Add(ctx context.Context, obj ApplyableObject) error
```
This interface needs to be heavily documented.

barney-s replied:
added initial docs. PTAL.

a-hilaly commented:
Thanks Barni for these changes! I have some high level thoughts on the PR:

1/ `ApplySet` package:
- It's really hard to tell what you changed from the upstream applyset implementation. Could we split this into multiple commits? Like first commit just copy-paste the original package, then another commit showing your modifications? Would make it way easier to review what actually changed.
- About the license - I'm not sure how this works, but since this is forked from kubectlcode, should it still say "The Kro Authors"?
- We need more docs on the structs/fields/methods. Maybe add a `doc.go` file explaining why we forked this instead of using an existing library and when we might unfork it?
- my last question is, looks like the KEP bringing `ApplySet` is and was alpha for multiple k8s versions. are there any plans of graduating it to Beta?

2/ This PR mixes the server siide apply implementation with other fixes like the "source-rgd" label bug. IMO it would be better to keep this PR focused on just `SSA/ApplySet`, and handle the label fix separately. Or at minimum, put them in different commits so we can track what changed for what reason. Is makes it easier to revert specific changes if something goes wrong.

3/ not needed in this PR, but we should track metrics for apply/prunes calls to the API server in a follow up PR

4/ RESTMapper: we already figure out the GVK and namespace for each resource during graph building. Do we really need the RESTMapper calls here? at the beginning of each reconciliation?

michaelhtm commented on pkg/applyset/prune.go:
```diff
@@ -0,0 +1,163 @@
+			results, err := a.findObjectsToPrune(ctx, dynamicClient, visitedUIDs, task.namespace, task.restMapping)
+			if err != nil {
+				return fmt.Errorf("listing %v objects for pruning: %w", task.restMapping.GroupVersionKind.String(), err)
+			} else {
+				task.results = results
+			}
```
nit: unnecessary else

michaelhtm commented on pkg/applyset/prune.go:
```diff
@@ -0,0 +1,163 @@
+
+// FindAllObjectsToPrune returns the list of objects that will be pruned.
+// Calling this instead of Prune can be useful for dry-run / diff behaviour.
+func (a *applySet) FindAllObjectsToPrune(
```
qq about this one..how do we know which objects we need to delete?
Let's say I remove one of the Resources from my RGD, do we also prune that here?

barney-s commented on pkg/controller/instance/instance_state.go:
```diff
@@ -48,3 +50,13 @@ type InstanceState struct {
 	ReconcileErr error
 }
+
+func (s *InstanceState) ResourceErrors() error {
```
`instance/controller_reconcile.go:246`

barney-s commented on pkg/controller/instance/controller.go:
```diff
@@ -154,8 +161,9 @@ func (c *Controller) Reconcile(ctx context.Context, req ctrl.Request) error {
 		runtime:                     rgRuntime,
-		instanceLabeler:             c.instanceLabeler,
+		instanceLabeler:             c.sourceLabeler,
 		instanceSubResourcesLabeler: instanceSubResourcesLabeler,
```
No this has to be different. Source labeller is specifically for just the instance object.

barney-s commented:
> 1/ `ApplySet` package:
> * It's really hard to tell what you changed from the upstream applyset implementation. Could we split this into multiple commits? Like first commit just copy-paste the original package, then another commit showing your modifications? Would make it way easier to review what actually changed.

I had to do a merge copy across two repos and simplify the code. Added the sources in docs.go. 
It was not a simple copy from just one repo. For me to split it into 2 commits, i would have to work back on the merge. 

> * About the license - I'm not sure how this works, but since this is forked from kubectlcode, should it still say "The Kro Authors"?

Since its not a simple copy over, we can add both authors for now.

> * We need more docs on the structs/fields/methods. Maybe add a `doc.go` file explaining why we forked this instead of using an existing library and when we might unfork it?

Added more comments (doc string) and a docs.go.

> * my last question is, looks like the KEP bringing `ApplySet` is and was alpha for multiple k8s versions. are there any plans of graduating it to Beta?

There is a comment on the KEP on what is needed to move it to Beta. At the moment it is not in progress. But we can keep pulling on that thread in k8s and see what can be done in parallel.

barney-s commented:
> 2/ This PR mixes the server siide apply implementation with other fixes like the "source-rgd" label bug. IMO it would be better to keep this PR focused on just `SSA/ApplySet`, and handle the label fix separately. Or at minimum, put them in different commits so we can track what changed for what reason. Is makes it easier to revert specific changes if something goes wrong.

separated out to https://github.com/kro-run/kro/pull/631

jakobmoellerdev commented on Makefile:
```diff
@@ -94,9 +94,9 @@ vet: ## Run go vet against code.
+test: manifests generate fmt vet envtest ## Run tests. Use WHAT=unit or WHAT=integration
 ifeq ($(WHAT),integration)
-	go test -v ./test/integration/suites/... -coverprofile integration-cover.out
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_VERSION) --bin-dir $(LOCALBIN) -p path)" go test -v ./test/integration/suites/... -coverprofile integration-cover.out
```
I just noticed this but aren't you basically changing the behavior of test to now implicitly always download envtest, and also run the integration tests always against envtest? 

I mean im not totally opposed to that but I was hoping that this would be a separate change as the specific CI step should be removed in that same notion: https://github.com/kro-run/kro/blob/main/.github/workflows/integration-tests.yaml#L31-L37

barney-s replied:
https://github.com/kro-run/kro/pull/636

barney-s commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,628 @@
+		parent.SetName("somename")
+
+	    // ApplySet configuration
+		applySetConfig := applyset.Config{
```
done.

barney-s commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,628 @@
+	Log          logr.Logger
+
+	// Callbacks
+	LoadCallback        func(obj ApplyableObject, clusterValue *unstructured.Unstructured) error
```
added.

barney-s commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,628 @@
+	DryRun(ctx context.Context, prune bool) (*ApplyResult, error)
+}
+
+type Config struct {
```
added

barney-s commented on pkg/controller/resourcegraphdefinition/controller_reconcile.go:
```diff
@@ -55,6 +58,7 @@ func (r *ResourceGraphDefinitionReconciler) reconcileResourceGraphDefinition(ctx
 	}
 
 	crd := processedRGD.Instance.GetCRD()
+	// TODO(barney-s): should we apply the source labeler to the crd here instead ?
```
removed. not part of this pr

barney-s commented on pkg/applyset/prune.go:
```diff
@@ -0,0 +1,163 @@
+	var tasks []*task
+
+	// We run discovery in parallel, in as many goroutines as priority and fairness will allow
+	// (We don't expect many requests in real-world scenarios - maybe tens, unlikely to be hundreds)
```
I do like the idea of limiting. GVKs in a single RGD could be bounded and not an issue with concurrency. 
That being said NumCPU may not be the appropriate limit since these are NW bound and not CPU bound tasks. 
Add a const instead ? Make that parameterizable later on.

barney-s replied:
set it to const 5

barney-s commented on pkg/applyset/prune.go:
```diff
@@ -0,0 +1,163 @@
+		restMapping *meta.RESTMapping
+		results     []PruneObject
+	}
+	var tasks []*task
```
done.

barney-s commented:
Gentle reminder... Been weeks now. PTAL @a-hilaly @matthchr

barney-s commented on applylib/applyset/prune.go:
```diff
@@ -0,0 +1,173 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package applyset
```
added tests

matthchr reviewed (approved):
LGTM. Had a few nits but overall happy with this.

barney-s commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,546 @@
+	Log          logr.Logger
+
+	// Callbacks
+	LoadCallback        func(obj ApplyableObject, clusterValue *unstructured.Unstructured) error
```
i have added some docs.

barney-s replied:
please check the latest code (comment was 2 weeks old)

barney-s commented on pkg/applyset/applyset.go:
```diff
@@ -0,0 +1,640 @@
+	return nil
+}
+
+func (a *applySet) validateAndCacheRestMapping(obj ApplyableObject) (*meta.RESTMapping, error) {
```
done

jakobmoellerdev commented on Makefile:
```diff
@@ -275,6 +275,7 @@ deploy-kind: ko
 	# This generates deployment with ko://... used in image.
 	# ko then intercepts it builds image, pushes to kind node, replaces the image in deployment and applies it
 	helm template kro ./helm --namespace kro-system --set image.pullPolicy=Never --set image.ko=true --set config.allowCRDDeletion=true | $(KO) apply -f -
+	sleep 5
```
This is a probably a leftover

jakobmoellerdev reviewed (commented):
**TL;DR**  
I would ask to vote in the next community call on whether to merge, then either follow up immediately or explore alternatives here. Sharing my viewpoints below for full transparency

Key critical review points for me:

1. **Callback mechanisms**  
   - Hard to debug: even if merged, we should define how to move away from them  
   - [Code reference](https://github.com/kro-run/kro/pull/561/files#diff-b51fa143abc50a3ef21525703711d6ab24b80395539ca3a6a9a4d3da24050435R191-R192)
   - Still [manual sync calls](https://github.com/kro-run/kro/pull/561/files#diff-b51fa143abc50a3ef21525703711d6ab24b80395539ca3a6a9a4d3da24050435R250) exist parallel to our normal sync loop → unclear code flow for me
   - Long-term maintainability IMO requires removing callbacks

3. **Testing**  
   - Consensus: more tests are required
   - Issue / Test Strategy unclear to me

4. **Applysets**  
   - `kubectl`’s applyset has flaws  
   - We may want to consider a “kro’ified” applyset aligned with reconcile loops

I really appreciate all the work done here, but there are (IMO) legit concerns that currently avoid maintainers to give a "full steam ahead", even though its likely the biggest feat for release

jakobmoellerdev commented:
How about this: 
- we do a quorum decision from all current maintainers if this should merge
- If we merge => post-merge iteration on main, blocking 0.5 until done (needs issues immediately so we can unblock 0.5 imho)
- If we do not merge => pre-merge on a separate branch
?

barney-s commented:
@a-hilaly - Would you please take a look and approve ?

a-hilaly commented on pkg/controller/instance/controller_reconcile.go:
```diff
@@ -123,193 +155,117 @@ func (igr *instanceGraphReconciler) reconcileInstance(ctx context.Context) error
-	descriptor := igr.runtime.ResourceDescriptor(resourceID)
-	gvr := descriptor.GetGroupVersionResource()
-	namespace := igr.getResourceNamespace(resourceID)
+	if _, err := igr.runtime.Synchronize(); err != nil {
+		return fmt.Errorf("failed to synchronize after apply/prune: %w", err)
+	}
```
Do we need to synchronize here? synchronize is only useful before applying new things.

a-hilaly commented on pkg/controller/instance/controller_reconcile.go:
```diff
@@ -123,193 +155,117 @@ func (igr *instanceGraphReconciler) reconcileInstance(ctx context.Context) error
-				resourceState.State = "WAITING_FOR_EXTERNAL_RESOURCE"
-				resourceState.Err = fmt.Errorf("external resource not found: %w", err)
-				return igr.delayedRequeue(resourceState.Err)
+		if clusterObj != nil {
```
will cluster can't be nil here.

barney-s replied:
it can be nil if it is applied for the first time. Because we havent yet applied. We just added to applyset. apply happens in 218.

a-hilaly commented on pkg/applyset/tracker.go:
```diff
@@ -0,0 +1,129 @@
+	// Lifecycle hints
+	// TODO(barney-s): need to exapnd on these: https://github.com/kro-run/kro/issues/542
+	ExternalRef bool
+	Decorate    bool
```
What is decorate used for?

a-hilaly reviewed (approved):
@barney-s Thank you for the persistence on this PR and addressing the feedback through multiple iterations. The work here is significant - implementing ApplySet for kro is a major architectural change and a positive one.

This is ready to merge. We can address any remaining items in follow-up PRs. Thank you again Barni! I've left some questions in the review below but these are minor compared to the overall value this brings.

a-hilaly commented:
@barney-s can you please squash the last 3 comments?

barney-s commented:
> @barney-s Thank you for the persistence on this PR and addressing the feedback through multiple iterations. The work here is significant - implementing ApplySet for kro is a major architectural change and a positive one.
>

Appreciate all the feedback and reviews. 
 
> This is ready to merge. We can address any remaining items in follow-up PRs. Thank you again Barni! I've left some questions in the review below but these are minor compared to the overall value this brings.

Addressed most of the comments/Qs.

a-hilaly reviewed (approved):
🚀 🚀 
/lgtm

allamand commented:
Nice work all 🚀
